ledger/participant-integration-api/src/main/resources/db/migration/postgres/V124__contract_keys.sql

```diff
@@ -0,0 +1,19 @@
+--  Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+--  SPDX-License-Identifier: Apache-2.0
+
+-- key lookups on create table not needed anymore anymore
+DROP INDEX participant_events_create_create_key_hash_idx;
+
+-- neither the hash in the create table
+ALTER TABLE participant_events_create DROP COLUMN create_key_hash;
+
+CREATE TABLE contract_keys (
+    contract_key_hash BIGINT NOT NULL,
+    create_event_sequential_id BIGINT NOT NULL,
+    contract_id TEXT NOT NULL
+);
+
+-- for key mapping lookup
+CREATE INDEX contract_keys_lookup ON contract_keys USING BTREE (contract_key_hash, create_event_sequential_id, contract_id);
+-- for pruning
+CREATE INDEX contract_keys_event_lookup ON contract_keys USING BTREE (contract_id);
```

Comment:
That one is brutal. A btree over a long uniformly distributed index. Why do we need that? Can't we lookup the event_seq_id through the create_events table, and trade a bit of read perf for write-perf?

Reply:
That is quite some read load, since for each archived contract we need to load the full heap page to get the even_seq_id.
And if we would decide to build a covering index there to not go to the heap, we would need to switch from a hash index to a very similar btree index.
Also it would cause similar problems in the MCSC: we would need a cache with two indexes (contract id and event_seq_id), which is additional effort and complexity (to a very complex code already). (or alternatively introduce another nasty index to lookup contract ids for event_seq_id).
-- That's why I was going for this at first, if this matches already the 5day ingestion, we got away easy :crossed_fingers:

Reply:
OK. Your call. I would argue that it's OK to keep contract data for contracts that were recently used in our page buffer. It will be present anyways for the submitter. I worry about this `contract_id` index here, and I was hoping we'd get away from btree's over uniformly distributed data.